Initialize buffer for `ConversationSummaryMemory` if underlying chat messages are not empty.

outday29 opened the pull request:
`ConversationSummaryMemory` will have an empty buffer even if the chat messages are not empty. This is because the first refresh of buffer occurs in `save_context`. 

```
from langchain.memory import ConversationSummaryMemory
from langchain.chains import ConversationChain
from langchain.chat_models import ChatOpenAI
from langchain.memory import ChatMessageHistory

# Assuming I have existing messages.
chat_history = ChatMessageHistory()
chat_history.add_user_message("Hello!")
chat_history.add_ai_message("Hey there, how may I assist you today?")
chat_history.add_user_message("Answer briefly. What are the first 3 colors of a rainbow?")
chat_history.add_ai_message("The first three colors of a rainbow are red, orange, and yellow.")

chat_model = ChatOpenAI(temperature=0)
memory = ConversationSummaryMemory(
    chat_memory=chat_history, 
    llm=ChatOpenAI(temperature=0) # <-- You need an LLM to summarize the conversation.
)
conversation = ConversationChain(
    llm=chat_model,
    memory=memory,
    verbose=True
)
conversation.run("And the next 4?")
```

Output, notice summary is empty:
```
> Entering new ConversationChain chain...
Prompt after formatting:
The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:

Human: And the next 4?
AI:

> Finished chain.
The next 4 what? Please provide more context so I can better understand your question.
```

outday29 committed:
Fix summary memory

outday29 committed:
lint

hwchase17 commented:
why only use last -2? shouldnt we want to summarize all?

we should do this properly.

1. add a validation assertion that if history is passed in, if there are messages then summary should not be empty
2. add a classmethod `from_messages` that can create a summary but looping over pairs of messages and iteratively summarizing them

outday29 committed:
improvement

outday29 commented:
Ah that was a glaring mistake.

Below are the changes:
- I have modified the function to summarize all messages. 
- I added a `summarize_step` argument that allows user to specify how many messages to summarize in one go. This may be helpful if the given message list is very long.
- User can also input their custom summary through `buffer` argument during initialization. This allow bypass of summarization.

As for the `from_message` class method, I am not quite sure how to implement that. We don't have access to the instance-specific summarization method during that time.

outday29 committed:
fix
